### PR TITLE
fix(rln-relay): graceful shutdown with non-zero exit code

### PR DESCRIPTION
**Pull request by rymnc**
# Description

This PR follows the pattern set by the waku_archive protocol, which propagates fatal error handling to appropriate protocols, without which the node goes into UB.

# Changes

- [x] new module `error_handling` in `waku_common`, which is imported by `waku_archive` and `waku_rln_relay`, exporting a type `OnFatalErrorHandler`
- [x] propagate `onFatalErrorAction` to `waku_rln_relay`

## How to test

1. Run a wakunode, after the protocols are mounted, disable your internet. you will see the following -

image

### CODE REVIEW

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim [hunk 1]
```diff
@@ -234,19 +238,19 @@ when defined(rln_v2):
     let membershipFee = g.membershipFee.get()
 
     var gasPrice: int
-    retryWrapper(gasPrice, RetryStrategy.new(), "Failed to get gas price"):
+    g.retryWrapper(gasPrice, RetryStrategy.new(), "Failed to get gas price"):
```
Comment (gabrielmer) [inline waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:241]: Shouldn't we delete here the `RetryStrategy.new()`?
Same in some other places in this file

**Review by gabrielmer — APPROVED** [review]
LGTM 🔥 🔥 🔥 

Thanks so much!

File: waku/waku_rln_relay/group_manager/on_chain/group_manager.nim (hunk 1 above)
Comment (rymnc) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim]: slipped thru the cracks, apologies - 8d27b961

Comment (rymnc) [reply waku/waku_rln_relay/group_manager/on_chain/group_manager.nim]: interesting that it compiles anyway :O

File: waku/waku_rln_relay/group_manager/on_chain/retry_wrapper.nim [hunk 2]
```diff
@@ -32,4 +34,8 @@ template retryWrapper*(res: auto,
       exceptionMessage = getCurrentExceptionMsg()
       await sleepAsync(retryStrategy.retryDelay)
   if shouldRetry:
-    raise newException(CatchableError, errStr & ": " & exceptionMessage)
+    if errCallback == nil:
+      raise newException(CatchableError, errStr & ": " & exceptionMessage)
```
Comment (Ivansete-status) [inline waku/waku_rln_relay/group_manager/on_chain/retry_wrapper.nim:38]: Shall we somehow state that the main reason is `errCallback == nil`?
```suggestion
      raise newException(CatchableError, errStr & " errCallback == nil: " & exceptionMessage)
```

File: waku/waku_rln_relay/rln_relay.nim [hunk 3]
```diff
@@ -33,16 +34,19 @@ import
...
+type 
+  WakuRlnConfig* = object
+    rlnRelayDynamic*: bool
+    rlnRelayCredIndex*: Option[uint]
+    rlnRelayEthContractAddress*: string
+    rlnRelayEthClientAddress*: string
+    rlnRelayCredPath*: string
+    rlnRelayCredPassword*: string
+    rlnRelayTreePath*: string
+    # defined in waku_archive, but do not want to import it here
```
Comment (Ivansete-status) [inline waku/waku_rln_relay/rln_relay.nim:46]: Sorry I can't quite understand the comment:)

**Review by Ivansete-status — APPROVED** [review]
Awesome! Thanks for it! :100:

Just added minor comments

File: waku/waku_rln_relay/rln_relay.nim (hunk 3 above)
Comment (rymnc) [reply waku/waku_rln_relay/rln_relay.nim]: addressed in efd1b017

**Review by NagyZoltanPeter — APPROVED** [review]
Great stuff! LGTM.

**Pull request merged**